Modularity features/unreleased variants trees

nphilipp opened the pull request:
Handle module state change messages -- retrieve MMD, create UnreleasedVariant, Tree objects in PDC.

nphilipp committed:
Import unreleased trees into PDC.

When receiving a message with a topic ending in 'module.state.change',
expect the payload to have 'state' (done/ready), 'koji_tag',
'module_uid', 'module_version', 'topdir'. The basename of 'topdir' is
the future tree ID of the form '<name>-<version>-<date>.<spin>',
containing one or more directories per architecture. List RPM files in
there and add them as trees to PDC.

nphilipp committed:
API end points are plural, not singular

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Add unit test for tree handler.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Add TmpDir, PushPopD context managers.

TmpDir creates a temporary directory when entering, and removes it and
its contents when exiting the context.

PushPopD changes into a directory when entering, and back into the
previous one when exiting the context.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Store module dependencies in PDC.

Rename trees -> modules to better reflect what is handled.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Filter out stdout of git commands.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Enhance unit tests for retrieving module metadata.

Test cloning a module repository, parsing the metadata YAML and adding
the dependency information to PDC.

Signed-off-by: Nils Philippsen <nils@redhat.com>

ralphbean commented on pdcupdater/handlers/modules.py:
```diff
@@ -0,0 +1,274 @@
+
+log = logging.getLogger(__name__)
+
+md_ns = {
```
What is this?

nphilipp replied:
@ralphbean, good catch! This is from a very early version of my handler that pulled info from repository metadata. I'll remove it.

ralphbean commented on pdcupdater/handlers/modules.py:
```diff
@@ -0,0 +1,274 @@
+            self.handle_new_tree(pdc, body, unreleased_variant)
+
+    def get_mmd_from_scm(self, scmurl):
+        with TmpDir(prefix="pdcupdater-") as tmpdir, PushPopD(tmpdir), \
```
I didn't know you could do comma-separated context managers!  This is great!

nphilipp replied:
I only found this while attempting to do it, too ;).

ralphbean commented on pdcupdater/handlers/modules.py:
```diff
@@ -0,0 +1,267 @@
+        state = msg['msg']['state']
+
+        if state not in self.valid_states:
+            log.warn("Invalid module state '{}', skipping.".format(state))
```
Maybe make this a `log.error` statement instead.

The reason is that (in the production environment) warnings are sent to the journal, but errors are sent to the journal and are also emailed to developers/sysadmins.  If we get a message with an invalid state, something is _very_ wrong and we want to know about it.

nphilipp replied:
Agreed.

ralphbean commented on pdcupdater/handlers/modules.py:
```diff
@@ -0,0 +1,267 @@
+
+        unreleased_variant = self.get_or_create_unreleased_variant(pdc, body)
+
+        if 'topdir' in body:
```
What scenario are we looking at if `'topdir'` is not in `body`?  Could you add an inline comment explaining this conditional for future developers?

nphilipp replied:
I understood that pdc-updater gets a message when a module is just submitted for building, i.e. no tree has been created yet. Before I add such a comment, does that sound about right? @lkocman, @contyk?

ralphbean replied:
That is true!  Yes.

You'll get messages for states 'init', 'wait', 'building', and 'failed'... in all those cases, you won't have a tree.

nphilipp replied:
Yeah, but the handler considers these states not relevant right now (and skips over). Should it?

nphilipp replied:
NB: Right now pdc-updater looks at states 'done' and 'ready'.

ralphbean commented on pdcupdater/handlers/modules.py:
```diff
@@ -0,0 +1,267 @@
+                open(os.devnull, "w") as devnull:
+            m = self.scmurl_re.match(scmurl)
+            if not m:
+                log.error("Can't parse SCM URL: {}".format(scmurl))
```
This log statement may not be necessary.

Exceptions raised here get caught at a higher level and logged, so we'd get two copies of this error.

nphilipp replied:
Ahh, I'll remove it then.

ralphbean commented on pdcupdater/handlers/modules.py:
```diff
@@ -0,0 +1,267 @@
+            repopath = m.group('repopath')
+            modpath = m.group('modpath')
+            revision = m.group('revision')
+            while repopath.endswith("/"):
```
You could do `repopath = repopath.rstrip('/')` here?

nphilipp replied:
Uhm, yes, I could. :)

ralphbean commented on pdcupdater/handlers/modules.py:
```diff
@@ -0,0 +1,267 @@
+            while repopath.endswith("/"):
+                repopath = repopath[:-1]
+            modname = repopath.rsplit("/", 1)[1]
+            if modname.endswith(".git"):
```
Same here.  `modname = modname.rstrip('.git')`.

nphilipp replied:
No, that would strip all trailing dots, g's, i's, t's ;).

ralphbean commented on pdcupdater/handlers/modules.py:
```diff
@@ -0,0 +1,267 @@
+                modname = modname[:-4]
+
+            log.debug("Cloning {}".format(giturl))
+            check_call(["git", "clone", "-n", giturl, modname],
```
Just for my information, what happens if `check_call` fails?

nphilipp replied:
It raises a `CalledProcessError` exception... Do you think it's better if we catch it, or even wrap it into our own exception? Handle the raise `CalledProcessError` further up in `handle()`?

ralphbean replied:
I'm indifferent.  If it raises an exception, that will get caught and logged.

Does the CalledProcessError have information about the command that it tried to run?  As long as it has that, that will be enough information to debug production issues in the future.

nphilipp replied:
The exception object will render something like this when printed/cast to str:

```
Command '['/bin/false']' returned non-zero exit status 1
```

Besides that, it has a `cmd` member (a list of the command and arguments).

ralphbean commented on pdcupdater/handlers/modules.py:
```diff
@@ -0,0 +1,267 @@
+            else:
+                build_deps.append(dep)
+
+        koji_tag = body['koji_tag']
```
I didn't realize the value of the `koji_tag` was supposed to come from the `module.state.change` message.

We keep it set to `None` when we first publish the message and were expecting pdc-updater to create it as something intelligent.

Can you replace this with some code that just "makes up" a good koji tag to use?  Just using "{module_name}-{module_version}-{module_release}" will work.

nphilipp replied:
Will do.

nphilipp replied:
Perhaps even something namespaced that makes it explicit that it's about a module being built, e.g. `module-<name>-<version>-<release`?

nphilipp replied:
And, right now I assume that `module_id` or `module_uid` contain name, version, release. Right or wrong?

ralphbean commented on pdcupdater/handlers/modules.py:
```diff
@@ -0,0 +1,267 @@
+
+        return unreleased_variant
+
+    def get_or_create_unreleased_variant(self, pdc, body):
```
I had trouble reading and understanding this function when I first encountered it.

Can you add some inline comments to it explaining the thought?

nphilipp replied:
I added a doc comment to the method.

ralphbean commented on pdcupdater/handlers/modules.py:
```diff
@@ -0,0 +1,267 @@
+        variant_id = variant_uid.lower()
+
+        try:
+            unreleased_variant = pdc['unreleasedvariants'][variant_id]._()
```
First we try and get it... right?

nphilipp replied:
Yup.

ralphbean commented on pdcupdater/handlers/modules.py:
```diff
@@ -0,0 +1,267 @@
+        try:
+            unreleased_variant = pdc['unreleasedvariants'][variant_id]._()
+        except beanbag.BeanBagException as e:
+            if e.response.status_code != 404:
```
And if it was a 404, then create it.

If it was some other error code, then bail out.  For instance, a 500 error means pdc is down and we can't proceed intelligently.

nphilipp replied:
Do you think we should have more commentary in the method for this?

ralphbean commented on pdcupdater/handlers/modules.py:
```diff
@@ -0,0 +1,267 @@
+            if e.response.status_code != 404:
+                raise
+        else:
+            log.info("Tree exists already, skipping: {}".format(tree_id))
```
Nice!

nphilipp replied:
NB: it doesn't check if the tree referenced in the message and the tree stored in PDC already are the same. Can you imagine a scenario where this might be the case, where we should check it?

ralphbean commented on pdcupdater/tests/handler_tests/test_modules.py:
```diff
@@ -0,0 +1,51 @@
+            'topic': 'buildsys.module.state.change',
+            'msg': {
+                'state': 'done',
+                'koji_tag': 'module-core-24',
```
Yeah, this is the result of some miscommunication I think.

When the module build is done, rida will have had to already know the koji_tag.

But when the module build is first starting... rida doesn't know it yet... and it was going to try and get it from pdc.  We had thought that the value was going to be synthesized and put in place by pdc-updater, but that's earlier on when the module yaml file is first committed to dist-git (I think).

nphilipp replied:
So pdc-updater should care about states other than 'done' and 'ready' after all :).

nphilipp replied:
Contrary to what I assumed previously, pdc-updater should process all `module.state.change` messages, and create `UnreleasedVariant` objects, containing e.g. dep information because that should be retrievable even in `init`, right? (ping @lkocman, @contyk)

nphilipp replied:
NB: Should pdc-updater do something (e.g. remove the associated `UnreleasedVariant`) if it gets a message with state `failed`?

nphilipp committed:
Add a doc comment to get_or_create_unreleased_variant().

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Document when we expect topdir/tree info in msg.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Process all non-failed module states.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Use simplified 'name', 'version', 'release' in the message.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Build variant_uid from name, version, release.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Create koji_tag ourselves.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
FOO: ac60cbe8973f93f9fad60379848042fa32d16b82

nphilipp committed:
Update unit tests for modules for recent changes.

Additionally, have two separate tests, one processing a message that a
module build process has been initiated, another for the followup
message when the build is finished.

Signed-off-by: Nils Philippsen <nils@redhat.com>
